# Pandas Final Assignment

In [1]:
import pandas as pd

So during this assignment we will study a dataset that includes house that were sold in year 2014 and 2015 at King County, US.
The main objective is to extract the maximum of variables that are affecting the prices in order to reach the ability to estimate the price of a new house knowing all of its characteristics (area, grade, view, location etc..)

#### Summary

The study that we will be doing will be based on the following main points:
- List all the variables that we have in our hand 
- Shortlist the variables that might potentially affect the prices
- Take out the average prices by grouping each of theses parameters separately
- Choose the columns that are proven to affect the price based on our analysis
- Find patterns and the degree of influence on each of these variables
- Find a price for a house with a missing price value based on all the previous study

#### Importing the dataset

In [2]:
df = pd.read_csv('kc_house_data.csv')

Let's start by taking a first look on our dataFrame

In [3]:
df.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0         3       1.00         1180   
1  6414100192  20141209T000000   538000.0         3       2.25         2570   
2  5631500400  20150225T000000   180000.0         2       1.00          770   
3  2487200875  20141209T000000   604000.0         4       3.00         1960   
4  1954400510  20150218T000000   510000.0         3       2.00         1680   
5  7237550310  20140512T000000  1230000.0         4       4.50         5420   
6  1321400060  20140627T000000   257500.0         3       2.25         1715   
7  2008000270  20150115T000000   291850.0         3       1.50         1060   
8  2414600126  20150415T000000   229500.0         3       1.00         1780   
9  3793500160  20150312T000000   323000.0         3       2.50         1890   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   
5    101930     1.0           0     0  ...     11        3890           1530   
6      6819     2.0           0     0  ...      7        1715              0   
7      9711     1.0           0     0  ...      7        1060              0   
8      7470     1.0           0     0  ...      7        1050            730   
9      6560     2.0           0     0  ...      7        1890              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   
5      2001             0    98053  47.6561 -122.005           4760   
6      1995             0    98003  47.3097 -122.327           2238   
7      1963             0    98198  47.4095 -122.315           1650   
8      1960             0    98146  47.5123 -122.337           1780   
9      2003             0    98038  47.3684 -122.031           2390   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  
5      101930  
6        6819  
7        9711  
8        8113  
9        7570  

[10 rows x 21 columns]

1- First, let's check all the variables available in our hand:

In [5]:
df.columns #List all the columns in our dataFrame

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

2- In order to simplify our anaysis, we will start by choosing some of these columns which potentially would affect the prices based on our first insights

The columns we will be choosing are the following:

Bedrooms, sqft_living, sqft_lot, floors, waterfont, view, grade, condition, yr_built, year renovated, lat, long

In [14]:
list_columns=['id','price','bedrooms', 'sqft_living', 'sqft_lot', 'floors', 
              'waterfront', 'view', 'grade', 'condition', 'yr_built','yr_renovated','lat', 'long'] #Save all the shortlisted column in a separate list

3- Now, we will group out the dataframe by each of these columns and check out the average prices:

3.1 First we start checking the bedrooms and living area size:

In [6]:
df[['price','bedrooms','sqft_living','sqft_lot']].groupby(['bedrooms']).mean().round(1).reset_index()

bedrooms      price  sqft_living  sqft_lot
0          0   410223.1       1707.8   24141.3
1          1   317658.0        884.6   16285.7
2          2   401387.7       1239.8   12096.7
3          3   466276.6       1805.8   14414.8
4          4   635564.7       2554.6   16720.8
5          5   786874.1       3047.5   16769.0
6          6   825853.5       3284.3   18586.9
7          7   951447.8       3974.2   20840.6
8          8  1105076.9       3800.0    8971.6
9          9   893999.8       3775.0    5480.0
10        10   820000.0       3706.7    8859.7
11        11   520000.0       3000.0    4960.0
12        33   640000.0       1620.0    6000.0

This table represents the average price and average living area size for each bedrooms number.
Looking at the results, as a first impression, we see that the average prices is going up in relation to the number of bedrooms except for certain cases like number 8 and number 9. However for this case the living area size is lower with 9 bedrooms. Which means that the living area size is more indicative in terms of influence on the prices. So for the following analysis we will be considering the sqft living area as the first indication to the prices due to the fact that the number of bedrooms is considered a consequence of the living area size.

However, for the sqft_lot size it does not seem to have such an impact on the prices as much as the living area. For example there are houses that have same lot area but different living area size and the difference in the price is noticeable.

3.2 We will do the same for the floors column: 

In [42]:
df[['price','floors','sqft_living']].groupby(['floors']).mean().round(1).reset_index()

floors      price  sqft_living
0     1.0   442219.6       1731.8
1     1.5   559044.9       1907.9
2     2.0   649051.5       2571.3
3     2.5  1061020.9       3145.4
4     3.0   582620.1       1788.3
5     3.5   933937.5       2518.0

In [7]:
df['floors'].value_counts()

floors
1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: count, dtype: int64

Looking at the results, most of the houses have between 1 and 2 floors. Despite the fact that the price is increasing with the number floors, we could notice that it is the sqft living area that is increasing as well so we will still be considering the living area as the first influencing variable

3.3: Group by condition

In [47]:
df[['price','condition','grade','sqft_living']].groupby(['condition']).agg({'price' : ['mean'], 'grade' : ['median'],'sqft_living':['mean']}).round(1).reset_index()

condition     price  grade sqft_living
                 mean median        mean
0         1  334431.7    6.0      1216.0
1         2  327316.2    7.0      1410.1
2         3  542097.1    8.0      2149.0
3         4  521300.7    7.0      1951.0
4         5  612577.7    7.0      2022.9

We will always be checking the value counts for each category to be sure wether our approach is being fair enough :

In [9]:
df['condition'].value_counts()

condition
3    14031
4     5679
5     1701
2      172
1       30
Name: count, dtype: int64

We displayed the average prices grouping by condition to check whether the condition has an influence on the price. Despite seeing an ascending pattern, we added up the grade and living area size to check wether it might be due to other variables. However, despite the fact that the living area sizes are changing, the condition still has an effect to a certain extent.

For example there is a difference between the average of prices for condition 4 and 5, although le grade and living area are almost the same.



3.4: Group by grade:

In [50]:
df[['price','grade']].groupby(['grade']).mean().round(1).reset_index()

grade      price
0       1   142000.0
1       3   205666.7
2       4   214381.0
3       5   248524.0
4       6   301916.6
5       7   402593.3
6       8   542895.5
7       9   773738.2
8      10  1072347.5
9      11  1497792.4
10     12  2192500.0
11     13  3710769.2

The grade variable is definitely one of the most important criterias on house pricings. As shown in the table, the pattern is very clear and average price is linearly increasing in relation to the grade

3.5: Group by year built:

In [53]:
df[['price','yr_built']].groupby(['yr_built']).mean().round(1).reset_index()

yr_built     price
0        1900  581536.6
1        1901  557108.3
2        1902  673192.6
3        1903  480958.2
4        1904  583867.8
..        ...       ...
111      2011  544648.4
112      2012  527437.0
113      2013  678599.6
114      2014  683792.7
115      2015  759970.9

[116 rows x 2 columns]

Clearly the price does not depend on how old the house is. As shown in the table, the average prices do not seem to be affected. However to be sure on this, there are houses that have been renovated and could have an effect on the price. In order to be sure we will have a look on how many houses have been renovated:

In [24]:
df['yr_renovated'].value_counts(normalize=True).mul(100)

yr_renovated
0       95.771064
2014     0.421043
2013     0.171193
2003     0.166566
2005     0.161940
          ...    
1951     0.004627
1959     0.004627
1948     0.004627
1954     0.004627
1944     0.004627
Name: proportion, Length: 70, dtype: float64

Only 5% of houses have been renovated. So the conclusion already made is proven to be correct.

3.6 Group by view and waterfront

In order not to have biased conclucions, we will add up the area size average to make sure that the area does not interfer on the prices.

In [74]:
df[['price', 'waterfront', 'view',
    'sqft_living', 'sqft_lot']].groupby(['waterfront','view']).mean().reset_index()

waterfront  view         price  sqft_living       sqft_lot
0           0     0  4.966235e+05  1997.761660   14179.237416
1           0     1  8.130549e+05  2570.861027   11891.347432
2           0     2  7.839551e+05  2649.277487   22158.072251
3           0     3  9.646851e+05  3039.201629   34123.415479
4           0     4  1.270713e+06  3412.619565   22720.282609
5           1     1  6.350000e+05  1940.000000  167125.000000
6           1     2  1.842188e+06  3369.125000   34060.000000
7           1     3  1.173605e+06  2485.263158   51986.000000
8           1     4  1.728300e+06  3268.133333   20061.251852

In [71]:
df['view'].value_counts()

view
0    19489
2      963
3      510
1      332
4      319
Name: count, dtype: int64

Most houses have 0 views. So we will do our analysis based on 0 or >0 views. 

Clearly there are differences on the prices.
As for houses with 1,2,3,4 views the prices do not seem to be affected.
For houses with same number of views, and with close area sizes, the waterfront variable does make a difference (Comparing row 4 and row 6).
So for the following analysis we will be based on the following:
- Whether a house has 0 or >0 views
- If the house has at least 1 view, the price will increase whether there is a waterfront or not.

In order to simplify our study, we will divide houe dataFrame into 2 categories: 

1- No View ('No')

2- With View ('Yes')


Let's implement it in our dataFrame:

In [15]:
df1=df[list_columns].copy()

In [17]:
df1['view']=df.apply(lambda x:'Yes' if x['view']>0  else 'No',axis=1)

In [18]:
df1.head()

id     price  bedrooms  sqft_living  sqft_lot  floors  waterfront  \
0  7129300520  221900.0         3         1180      5650     1.0           0   
1  6414100192  538000.0         3         2570      7242     2.0           0   
2  5631500400  180000.0         2          770     10000     1.0           0   
3  2487200875  604000.0         4         1960      5000     1.0           0   
4  1954400510  510000.0         3         1680      8080     1.0           0   

  view  grade  condition  yr_built  yr_renovated      lat     long  
0   No      7          3      1955             0  47.5112 -122.257  
1   No      7          3      1951          1991  47.7210 -122.319  
2   No      6          3      1933             0  47.7379 -122.233  
3   No      7          5      1965             0  47.5208 -122.393  
4   No      8          3      1987             0  47.6168 -122.045

3.7 Location:Long and Lat:

 We know for a fact that the location does have an influence on the price, so we will keep it in the updated list of columns for further analysis.

4- Based on our analysis, we will pick up all the variables that are proven to be affecting the price. Thus, the list of columns will have to be updated:

In [19]:
list_columns=['id','price','sqft_living','waterfront', 'view', 'grade', 'condition','lat', 'long'] #Updated list of columns

In [20]:
df1=df1[list_columns] 

In [21]:
df1.head()

id     price  sqft_living  waterfront view  grade  condition  \
0  7129300520  221900.0         1180           0   No      7          3   
1  6414100192  538000.0         2570           0   No      7          3   
2  5631500400  180000.0          770           0   No      6          3   
3  2487200875  604000.0         1960           0   No      7          5   
4  1954400510  510000.0         1680           0   No      8          3   

       lat     long  
0  47.5112 -122.257  
1  47.7210 -122.319  
2  47.7379 -122.233  
3  47.5208 -122.393  
4  47.6168 -122.045

5- Geographical Location:

In order to figure out the prices of the houses with the variables we have, we are required to specify a fixed area on the map since prices could vary largely depending on the location of the house

We will add the distance function given 2 latitudes and longitudes 

In [22]:
from math import sin, cos, sqrt, atan2
def distance(lat1,lon1,lat2,lon2):
    R = 6373.0
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = (sin(dlat/2))**2 + cos(lat1) * cos(lat2) * (sin(dlon/2))**2
    c = 2 * atan2(sqrt(a), sqrt(1-a))
    distance = R * c
    return distance

We will have a on the location where all the houses are located by finding the min and max longitudes and latitudes:

In [23]:
df1[['long','lat']].agg(['min','max'])

long      lat
min -122.519  47.1559
max -121.315  47.7776

#### Example 1

We will choose a randow sqare area with a ~ 30 km diameter:
- Longitude between -121.850 and -121.845
- Latitude between 47.5290 and 47.5330

In [24]:
distance(47.4880,-122.845,47.4880,-122.850)

29.7759587284091

In [25]:
distance(47.5290,-122.780,47.5330,-122.780)

25.49199999998587

Let us show a subset of all the houses located in this area in order to have a look on the prices: We will insert the results in a separate dataFrame named 'df_area1'

In [28]:
df_area1=df1.loc[(df1['long'] > -121.850) & (df1['long'] <-121.845) & (df1['lat'] >47.5290) & (df1['lat'] <47.5330)]

In [29]:
df_area1.sort_values('price', ascending =True)

id     price  sqft_living  waterfront view  grade  condition  \
9255   7853230720  368000.0         2080           0   No      7          3   
4602   7853230590  435000.0         2190           0   No      7          3   
11976  7853230570  440000.0         2230           0   No      7          3   
10760  7853230200  470000.0         2480           0   No      7          3   
16150  7853230460  555000.0         2690           0   No      7          3   

           lat     long  
9255   47.5300 -121.848  
4602   47.5305 -121.847  
11976  47.5308 -121.847  
10760  47.5297 -121.846  
16150  47.5302 -121.849

- Now we have a dataframe of 5 houses with the same view, grade, condition and waterfront.
- These houses are located within the same area.
- By sorting the dataFrame by price in the ascending order, it is very clear that the prices are increasing proportionally with the sqft_living area.

By obtaining these results we can now estimate the prices of the houses located within this area by calculating the price per sqft living:

In [30]:
df_area1['price/sqft']=df_area1['price']/df_area1['sqft_living']

C:\Users\lenovo\AppData\Local\Temp\ipykernel_7508\504626660.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_area1['price/sqft']=df_area1['price']/df_area1['sqft_living']


In [31]:
df_area1

id     price  sqft_living  waterfront view  grade  condition  \
4602   7853230590  435000.0         2190           0   No      7          3   
9255   7853230720  368000.0         2080           0   No      7          3   
10760  7853230200  470000.0         2480           0   No      7          3   
11976  7853230570  440000.0         2230           0   No      7          3   
16150  7853230460  555000.0         2690           0   No      7          3   

           lat     long  price/sqft  
4602   47.5305 -121.847  198.630137  
9255   47.5300 -121.848  176.923077  
10760  47.5297 -121.846  189.516129  
11976  47.5308 -121.847  197.309417  
16150  47.5302 -121.849  206.319703

In [32]:
df_area1[['price/sqft']].agg(['min','max','mean'])

price/sqft
min   176.923077
max   206.319703
mean  193.739693

So the prices of the houses located within the fixed area with 0 view, 0 waterfront, grade 7 are between 176$ and 206$ which gives us an average of 193$/sqft 

6- Estimation of prices 

In the final step, we will consider that we have a missing price for a certain house and we want to estimate the fair price based on our previous analysis. 
We will take a random price value from the df_area2 dataFrame and we will insert a missing value instead

In [37]:
import numpy as np
missing_percentage = 0.1  # 10% missing values

In [46]:
df_missing = df_area1.copy()

In [49]:
mask = np.random.rand(*df_missing['price'].shape)<missing_percentage
df_missing['price'][mask] = np.nan

C:\Users\lenovo\AppData\Local\Temp\ipykernel_7508\1215590272.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_missing['price'][mask] = np.nan


In [50]:
df_missing

id     price  sqft_living  waterfront view  grade  condition  \
4602   7853230590  435000.0         2190           0   No      7          3   
9255   7853230720  368000.0         2080           0   No      7          3   
10760  7853230200       NaN         2480           0   No      7          3   
11976  7853230570  440000.0         2230           0   No      7          3   
16150  7853230460  555000.0         2690           0   No      7          3   

           lat     long  price/sqft  
4602   47.5305 -121.847  198.630137  
9255   47.5300 -121.848  176.923077  
10760  47.5297 -121.846  189.516129  
11976  47.5308 -121.847  197.309417  
16150  47.5302 -121.849  206.319703

So now we have a missing price on the ID no. 7853230200 with a sqft_living=2480.
Following our study, we have to fill in the price as equal to sqft_living * average price/sqft: 2480 * 193

In [51]:
avg_price_sqft=193 # The average price per sqft

In [57]:
df_filled=df_missing.copy()

In [58]:
df_filled['price'] = df_missing['price'].fillna(df_missing['sqft_living']*avg_price_sqft)

In [59]:
df_filled

id     price  sqft_living  waterfront view  grade  condition  \
4602   7853230590  435000.0         2190           0   No      7          3   
9255   7853230720  368000.0         2080           0   No      7          3   
10760  7853230200  478640.0         2480           0   No      7          3   
11976  7853230570  440000.0         2230           0   No      7          3   
16150  7853230460  555000.0         2690           0   No      7          3   

           lat     long  price/sqft  
4602   47.5305 -121.847  198.630137  
9255   47.5300 -121.848  176.923077  
10760  47.5297 -121.846  189.516129  
11976  47.5308 -121.847  197.309417  
16150  47.5302 -121.849  206.319703

As shown in the table, the filled value is 478640 which is an estimation of the price based on the living area.
The initial price before inserting the missing value was 470000 which gives an acceptable estimation with 2% margin of error.

#### Example 2:

In this example we will choose another area defined by a rectangle between the 2 following coordinates:
- Longitude between -122.405 and -122.400
- Latitude between 47.6370 and 47.6400


In [60]:
distance(47.6400,-122.405,47.6370,-122.405)

19.119000000000725

In [61]:
distance(47.6400,-122.400,47.6400,-122.405)

27.714429368377395

The area chosen has 27 km between longitudes and 19 km between latitudes

In [77]:
df_area2=df1.loc[(df['long'] > -122.405) & (df['long'] <-122.400) & (df['lat'] >47.6370) & (df['lat'] <47.6400)]

In [78]:
df_area2

id      price  sqft_living  waterfront view  grade  condition  \
1962   1370802650   605000.0         2660           0   No      7          4   
4524   1370802540   875000.0         2720           0  Yes      8          4   
4633   1370802600  1020000.0         3620           0   No     10          3   
7708   1370802620   592500.0         1340           0   No      8          3   
18072  1370803835   705000.0         2320           0   No      8          5   
18516  1370802770   849000.0         2520           0   No      9          5   
19570  1370803640   619790.0         1040           0   No      7          4   

           lat     long  
1962   47.6391 -122.403  
4524   47.6384 -122.404  
4633   47.6393 -122.403  
7708   47.6384 -122.403  
18072  47.6398 -122.403  
18516  47.6381 -122.401  
19570  47.6385 -122.401

So this area includes a dataFrame of 7 houses:

The remarkable thing about this area is that it has a variation of grades (between 7 and 10) and condition (between 3 and 5) which is different than the case in the previous area. So in this study we will have the chance to have a look on how these variables are affecting the prices.

Since the grade variable seems like the most influencing we will start by sorting our dataFrame by grade:

In [79]:
df_area2.sort_values('grade', ascending =True,inplace=True)

C:\Users\lenovo\AppData\Local\Temp\ipykernel_7508\3851053812.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_area2.sort_values('grade', ascending =True,inplace=True)


Then we will add a price/sqft column to analyze the prices and see how the grade and contion are affecting them

In [80]:
df_area2['price/sqft']=df_area2['price']/df_area2['sqft_living']

C:\Users\lenovo\AppData\Local\Temp\ipykernel_7508\3728295466.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_area2['price/sqft']=df_area2['price']/df_area2['sqft_living']


In [81]:
df_area2

id      price  sqft_living  waterfront view  grade  condition  \
1962   1370802650   605000.0         2660           0   No      7          4   
19570  1370803640   619790.0         1040           0   No      7          4   
4524   1370802540   875000.0         2720           0  Yes      8          4   
7708   1370802620   592500.0         1340           0   No      8          3   
18072  1370803835   705000.0         2320           0   No      8          5   
18516  1370802770   849000.0         2520           0   No      9          5   
4633   1370802600  1020000.0         3620           0   No     10          3   

           lat     long  price/sqft  
1962   47.6391 -122.403  227.443609  
19570  47.6385 -122.401  595.951923  
4524   47.6384 -122.404  321.691176  
7708   47.6384 -122.403  442.164179  
18072  47.6398 -122.403  303.879310  
18516  47.6381 -122.401  336.904762  
4633   47.6393 -122.403  281.767956

Now we have a table with the average price of each house and sorted by grade
From this table we can draw few conclusion by taking some examples:

1- The first two houses have the same grade and condition, but there is a very high difference in the price rate (227$/sqft and 595$/sqft).
This shows that despite the fact that both houses are within a same small area, there is an area which is much more expensive than the other. For example the house at index 19570 could located in a downtown or at the center of the city. However the house at index 1962 is located in the suburbs


2- If we take the house at index 18072 (grade 8) and the house at index (grade 9) :
They are probably located within the same area where the price rates are the same. However the grade of the house (from grade 8 to grade 9) can increase the prices by 10%. (303$/sqft and 336$/sqft)

3- If we take the last 2 houses on the table, the difference (336$/sqft and 281$/sqft) might be due to the condition impact ( 3 and 5)

#### Conclusion based on the houses IDs:

Looking at both examples above, it is clear that there are a correlation between the IDs and the geographical location:

- Example 1: the houses located within the area1 have the same ID prefix which is #7853230XXX.
- Example 2: the houses located within the area2 have the same ID prefix which is #137080XXXX.

Based on that, the house ID could be an important indicator to the house location. However, if we take the second example we can notice that whenever the IDs difference is higher than 200-300 the house could be located on a totally different area within the same small rectangle



#### Example 3 based on IDs:

We will define a new area based on the houses IDs:

In [73]:
df1[['id']].agg(['min','max','median'])

id
min     1.000102e+06
max     9.900000e+09
median  3.904930e+09

The houses IDs are between 1000102 and 9900000000

In [136]:
df_area3=df1.loc[(df1['id'] > 3905100200) & (df1['id'] <3905939800)]

In [140]:
df_area3.sort_values('id', ascending =True,inplace=True)

C:\Users\lenovo\AppData\Local\Temp\ipykernel_7508\568318992.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_area3.sort_values('id', ascending =True,inplace=True)


In [141]:
df_area3['price/sqft']=df_area3['price']/df_area3['sqft_living']

C:\Users\lenovo\AppData\Local\Temp\ipykernel_7508\3161188995.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_area3['price/sqft']=df_area3['price']/df_area3['sqft_living']


In [142]:
df_area3

id     price  sqft_living  waterfront view  grade  condition  \
7687   3905100210  449950.0         1530           0   No      8          3   
7559   3905100220  535000.0         1720           0   No      8          3   
5179   3905100280  478000.0         1640           0   No      8          3   
12449  3905100310  544000.0         2030           0   No      8          3   
5264   3905100380  535000.0         1860           0   No      8          3   
1146   3905100520  510000.0         1860           0   No      8          3   
14223  3905100610  517500.0         1520           0   No      8          3   
17149  3905100630  500000.0         1710           0   No      8          4   
4690   3905100740  540000.0         1780           0   No      8          3   
17545  3905100840  500000.0         1580           0   No      8          4   
12877  3905120330  575000.0         2040           0   No      8          4   
12799  3905120540  570000.0         2290           0   No      8          3   
19518  3905120610  578000.0         2070           0   No      8          3   
2203   3905120830  612000.0         2180           0   No      8          3   

           lat     long  price/sqft  
7687   47.5691 -122.007  294.084967  
7559   47.5690 -122.007  311.046512  
5179   47.5689 -122.006  291.463415  
12449  47.5692 -122.006  267.980296  
5264   47.5699 -122.006  287.634409  
1146   47.5703 -122.004  274.193548  
14223  47.5696 -122.004  340.460526  
17149  47.5691 -122.004  292.397661  
4690   47.5695 -122.006  303.370787  
17545  47.5694 -122.005  316.455696  
12877  47.5719 -122.007  281.862745  
12799  47.5714 -122.005  248.908297  
19518  47.5706 -122.006  279.227053  
2203   47.5723 -122.007  280.733945

In [143]:
df_area3['price/sqft'].agg(['min','max','mean'])

min     248.908297
max     340.460526
mean    290.701418
Name: price/sqft, dtype: float64

The price of the houses are more or less stable depending on the area. The house condition is playing a slight role on the prices .

#### Price binning

In order to simplify our analysis it is better to distribute the price into different categories: A,B,C,D from the most to the least expensive

In [15]:
def condition(x):
    if (x>=75000 and x<200000): 
        return 'E' 
    elif (x>=200000 and x<400000): 
        return 'D' 
    elif (x>=400000 and x<800000): 
        return 'C'
    elif (x>=800000 and x<1200000):
        return 'B'
    else:
         return 'A'

In [16]:
df2=df.copy()

In [17]:
df2['Price_Category']=df2['price'].apply(condition)

In [18]:
df2['Price_Category'].value_counts(normalize=True).mul(100)

Price_Category
C    45.884421
D    36.589090
B     9.175034
A     4.714755
E     3.636700
Name: proportion, dtype: float64

In [19]:
df2[['price','bedrooms','sqft_living','grade','condition','sqft_basement','Price_Category']].groupby('Price_Category').agg({ 'sqft_living' :'mean','grade':'median','condition':'median','bedrooms':'median','sqft_basement':'mean'}).reset_index()

Price_Category  sqft_living  grade  condition  bedrooms  sqft_basement
0              A  3998.728165   10.0        3.0       4.0     744.500491
1              B  3051.880484    9.0        3.0       4.0     449.961674
2              C  2130.231522    8.0        3.0       3.0     319.023798
3              D  1621.384421    7.0        3.0       3.0     182.526050
4              E  1118.162850    6.0        3.0       3.0      53.804071

This table shows that both the living area and the grade are directly related and have the biggest impact on the prices, along with the location.However regarding the condition, it does no interfere as much regarding the price catogeries but could still have a small impact on the houses where the prices are relatively close

Example 3:

Let's have a look and some of the houses sorted by ID:

In [20]:
df_id=df.sort_values('id', ascending =True)

In [23]:
df_id.head(15)

id             date     price  bedrooms  bathrooms  sqft_living  \
2497   1000102  20150422T000000  300000.0         6       3.00         2400   
2496   1000102  20140916T000000  280000.0         6       3.00         2400   
6735   1200019  20140508T000000  647500.0         4       1.75         2060   
8411   1200021  20140811T000000  400000.0         3       1.00         1460   
8809   2800031  20150401T000000  235000.0         3       1.00         1430   
3557   3600057  20150319T000000  402500.0         4       2.00         1650   
18521  3600072  20150330T000000  680000.0         4       2.75         2220   
3200   3800008  20150224T000000  178000.0         5       1.50         1990   
21063  5200087  20140709T000000  487000.0         4       2.50         2540   
4333   6200017  20141112T000000  281000.0         3       1.00         1340   
16715  7200080  20141104T000000  239000.0         4       2.00         1980   
16814  7200179  20141016T000000  150000.0         2       1.00          840   
16815  7200179  20150424T000000  175000.0         2       1.00          840   
8508   7400062  20140521T000000  299800.0         2       1.00          790   
4481   7600057  20140805T000000  520000.0         3       2.00         1410   

       sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
2497       9373     2.0           0     0  ...      7        2400   
2496       9373     2.0           0     0  ...      7        2400   
6735      26036     1.0           0     0  ...      8        1160   
8411      43000     1.0           0     0  ...      7        1460   
8809       7599     1.5           0     0  ...      6        1010   
3557       3504     1.0           0     0  ...      7         760   
18521      5310     1.0           0     0  ...      7        1170   
3200      18200     1.0           0     0  ...      7        1990   
21063      5001     2.0           0     0  ...      9        2540   
4333      21336     1.5           0     0  ...      5        1340   
16715     10585     1.5           0     0  ...      6        1980   
16814     12750     1.0           0     0  ...      6         840   
16815     12750     1.0           0     0  ...      6         840   
8508       5240     1.0           0     0  ...      6         790   
4481       2700     2.0           0     0  ...      7        1410   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
2497               0      1991             0    98002  47.3262 -122.214   
2496               0      1991             0    98002  47.3262 -122.214   
6735             900      1947             0    98166  47.4444 -122.351   
8411               0      1952             0    98166  47.4434 -122.347   
8809             420      1930             0    98168  47.4783 -122.265   
3557             890      1951          2013    98144  47.5803 -122.294   
18521           1050      1951             0    98144  47.5801 -122.294   
3200               0      1960             0    98178  47.4938 -122.262   
21063              0      2005             0    98108  47.5423 -122.302   
4333               0      1945             0    98032  47.4023 -122.273   
16715              0      1924             0    98055  47.4836 -122.214   
16814              0      1925             0    98055  47.4840 -122.211   
16815              0      1925             0    98055  47.4840 -122.211   
8508               0      1925             0    98118  47.5303 -122.288   
4481               0      1902             0    98122  47.6029 -122.302   

       sqft_living15  sqft_lot15  
2497            2060        7316  
2496            2060        7316  
6735            2590       21891  
8411            2250       20023  
8809            1290       10320  
3557            1480        3504  
18521           1540        4200  
3200            1860        8658  
21063           2360        6834  
4333            1340       37703  
16715           1360        7810  
16814           1480 

Looking at this table, we can draw 2 main conclusions:

1- There are rows with same IDs which means that it is the same house sold 2 times, once each year. Based on this we can take some examples and have an idea about the price increase between 2014 and 2015:

Example 1: ID no.1000102:
This house was sold in 2014 at 280000$ then in 2015 at 300000$ which give a 7% increase.

Example 2: ID no.7200179:
This house was sold in 2014 at 150000$ then in 2015 at 175000$ which give a 8.5% increase.

Following the same logic we can take a sample of rows with same house IDs and calculate the average of the price increase between 2014 and 2015

### Conclusion

During this assignment, we had a target which was to find the main variables that have a direct effect on the prices in order to estimate a missing price value in the dataFrame.

Based on our study we found a list of columns that when grouping by could show a clear pattern on the prices, so we selected these variables and inserted them in a list named list_columns

Then we took a sample of 3 different areas by precising a square/rectange with a small diameter and extracted the houses within these areas. After then we calculated the average price/sqft and compared these values by looking at the other variables that might affect the prices. 

The first area we chose, had the same variables and the average of prices /sqft seemed stable. So we were able to insert missing values and filled them with the average price found and proved to be more or less accurate 

Finally, we could notice from sorting the houses by ID that there is a pattern between the IDs and the geographical location of the houses. In addition to that, there are houses that were sold twice between 2014 and 2015 and based on that we could get an idea about the price increase between those 2 years.